third_party/move/move-vm/types/src/value_serde.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::values::{DeserializationSeed, SerializationReadyValue, SizedID, Struct, Value};
+use move_binary_format::errors::PartialVMResult;
+use move_core_types::value::{IdentifierMappingKind, MoveTypeLayout};
+use serde::{
+    de::{DeserializeSeed, Error as DeError},
+    ser::Error as SerError,
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::str::FromStr;
+
+pub trait CustomDeserialize {
+    fn custom_deserialize<'d, D: Deserializer<'d>>(
+        &self,
+        deserializer: D,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+    ) -> Result<Value, D::Error>;
+}
+
+pub trait CustomSerialize {
+    fn custom_serialize<S: Serializer>(
+        &self,
+        serializer: S,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+        sized_id: SizedID,
+    ) -> Result<S::Ok, S::Error>;
+}
+
+// FIXME: consolidate
+fn is_string_layout(layout: &MoveTypeLayout) -> bool {
+    use MoveTypeLayout::*;
+    if let Struct(move_struct) = layout {
+        if let [Vector(elem)] = move_struct.fields().iter().as_slice() {
+            if let U8 = elem.as_ref() {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+fn is_derived_string_struct_layout(layout: &MoveTypeLayout) -> bool {
+    use MoveTypeLayout::*;
+    if let Struct(move_struct) = layout {
+        if let [value_field, Vector(padding_elem)] = move_struct.fields().iter().as_slice() {
```

Comment:
How about 
```
if let [value_field, padding_field] = ... {
   if is_string_layout(value_field) && is_string_layout(padding_field) {
     return true;
   }
 }
```

Reply:
I don't think you can do this because we have derived string snapshot as
```
struct DerivedStringSnapshot has store, drop {
  value: String,
  padding: vector<u8>,
}
```